tools/block-generator/generator/generate.go

```diff
@@ -763,110 +876,242 @@ func (g *generator) generateAssetTxnInternalHint(txType TxTypeID, round uint64,
+
+	if appTxType == appTxTypeOptin {
+		// pick a random app to optin:
+		appID := g.appSlice[kind][rand.Uint64()%numAppsForKind].appID
+
+		_, exists, optedIn := g.getAppData(true /* existing */, kind, senderIndex, appID)
+		if !exists {
+			return txType, kind, appTxType, appID, fmt.Errorf("should never happen! app %d of kind %s does not exist", appID, kind)
 		}
 
-		txn = g.makeAppCreateTxn(senderAcct, round, intra, approval, clear)
+		if optedIn {
+			// already optedin, so call the app instead:
+			return getAppTxType(kind, appTxTypeCall), kind, appTxTypeCall, appID, nil
+		}
 
-		appID := g.txnCounter + intra + 1
-		holding := &appHolding{appIndex: appID}
-		ad := &appData{
-			appID:    appID,
-			creator:  senderIndex,
-			kind:     kind,
-			holdings: []*appHolding{holding},
-			holders:  map[uint64]*appHolding{senderIndex: holding},
+		_, _, optedInPending := g.getAppData(false /* pending */, kind, senderIndex, appID)
+		if optedInPending {
+			// about to get opted in, but can't optin twice or call yet, so create:
+			return getAppTxType(kind, appTxTypeCreate), kind, appTxTypeCreate, appID, nil
 		}
-		g.pendingApps[kind] = append(g.pendingApps[kind], ad)
+		// not opted in or pending, so optin:
+		return txType, kind, appTxType, appID, nil
 	}
 
-	// account := indexToAccount(senderIndex)
-	// txn = g.makeAppCallTxn(account, round, intra, round, approval, clear)
+	if appTxType != appTxTypeCall {
+		return "", 0, 0, 0, fmt.Errorf("unimplemented transaction type for app %s from %s", appTxType, txType)
+	}
+	// WLOG appTxTypeCall:
 
-	if g.balances[senderIndex] < g.params.MinTxnFee {
-		return "", transactions.Transaction{}, fmt.Errorf("the sender account does not have enough algos for the app call. idx %d, app transaction type %v, num %d\n\n", senderIndex, txType, g.reportData[txType].GenerationCount)
+	numAppsOptedin := uint64(len(g.accountAppOptins[kind][senderIndex]))
+	if numAppsOptedin == 0 {
+		// try again calling recursively but attempting to optin:
+		return g.getActualAppCall(getAppTxType(kind, appTxTypeOptin), senderIndex)
 	}
-	g.balances[senderIndex] -= g.params.MinTxnFee
+	// WLOG appTxTypeCall with available optins:
 
-	return actual, txn, nil
+	appID := g.accountAppOptins[kind][senderIndex][rand.Uint64()%numAppsOptedin]
+	return txType, kind, appTxType, appID, nil
 }
 
-// ---- miscellaneous ----
+// ---- metric data recorders ----
 
 func track(id TxTypeID) (TxTypeID, time.Time) {
 	return id, time.Now()
 }
 
 func (g *generator) recordData(id TxTypeID, start time.Time) {
+	g.recordOccurrences(id, 1, start)
+}
+
+func (g *generator) recordOccurrences(id TxTypeID, count uint64, start time.Time) {
+	g.latestData[id] += count
 	data := g.reportData[id]
-	data.GenerationCount++
+	data.GenerationCount += count
 	data.GenerationTime += time.Since(start)
 	g.reportData[id] = data
 }
 
+func (g *generator) countAndRecordEffects(id TxTypeID, start time.Time) (uint64, error) {
+	g.recordData(id, start) // this may be a bug!!!
```

Comment:
Looks like a bug. You called this in `generateAppTxn`, so the generation time is already covered. Also... where does start come from?

Reply:
Wouldn't the inner transactions would be a side effect of the top level transaction instead of something the block generator has to worry about? This effects stuff seems unnecessary.

Reply:
> Looks like a bug ...
Good catch. I created a code smell and put a comment to remind myself to look more into it. Either effects should be removed from the report (which calls into question the need for effects), or at least they should keep `data.GenerationTime` zero-valued so that if at we want to subtract out the generation time we can do so accurately (this is introducing an _overcount_)

> Wouldn't the inner transactions ...
If conduit can be relied on counting the transactions correctly then I agree that counting the `effects` is superfluous. However, knowing how many transactions are expected is a good sanity check.

I think I see a better way forward. `effects` are still useful for reporting purposes. However, the runner should simply apply the effects map to the generated report and then add those statistics to the printout.

Reply:
Except one caveat - I still need to move the `txnCounter` forward correctly, and calculating the number of transactions directly from the effects is still better than repeating the calculation in a separate map. 

Summarizing what I'm planning to do:

- [ ] the values of the `effects` map should be slices of a type different from `TxTypeID`
- [ ] the generator only uses `effects` map to update `txnCounter` (or `intra`)
- [ ] there is a unit test that asserts the total number of transactions for each `TxTypeID` with effects is exactly as expected
- [ ] the effects are only used at report time in `run.go` to split out the various effects

Reply:
resolving as moved the TODO's into the main PR description and this is being addressed in a follow up PR with commit  https://github.com/tzaffi/go-algorand/pull/24/commits/291d30fe472810a195c2ccdd12be70ebd213b6b7